File: Strategy.FSharp/Hexagon.fs
module Strategy.FSharp.Hexagon

open System
open Godot


let CELL_SIZE = 40.0f

let CubeRound x y z =
    let mutable rx = round x
    let mutable ry = round y
    let mutable rz = round z

    let xDiff = abs <| rx - x
    let yDiff = abs <| ry - y
    let zDiff = abs <| rz - z

    if xDiff > yDiff && xDiff > zDiff then
        rx <- -ry - rz
    else if yDiff > zDiff then
        ry <- -rx - rz
    else
        rz <- -rx - ry

    (int32 rx, int32 ry, int32 rz)

let CalculateAxis axis1 axis2 = -axis1 - axis2

type Direction =
    | East = 0
    | NorthEast = 1
    | NorthWest = 2
    | West = 3
    | SouthWest = 4
    | SouthEast = 5

type Hexagon =
    { Q: int32
      R: int32
      S: int32 }

    static member Zero = { Q = 0; R = 0; S = 0 }

    static member NewAxial q r =
        // https://www.redblobgames.com/grids/hexagons/#conversions-axial
        { Q = q; R = r; S = CalculateAxis q r }

    static member NewCube q r s = { Q = q; R = r; S = s }

    static member At2DPosition(pos: Vector2) =
        let q =
            ((sqrt <| 3f) / 3f * pos.X - 1f / 3f * pos.Y)
            / CELL_SIZE

        let r = (2f / 3f * pos.Y) / CELL_SIZE
        let s = -q - r
        let q, r, s = CubeRound q r s
        Hexagon.NewCube q r s

    member this.Get2DPosition =
        let x =
            CELL_SIZE
            * (sqrt 3f * float32 this.Q
               + sqrt 3f / 2f * float32 this.R)

        let y = CELL_SIZE * (3f / 2f * float32 this.R)

        Vector2(x, y)

    // Create from Vector2 representation for easy passing to and from Godot
    static member FromVector2(vector: Vector2) =
        let q = int32 vector.X
        let r = int32 vector.Y
        Hexagon.NewAxial q r

    // Represent as Vector2 for easy passing to and from Godot
    member this.AsVector2 =
        let x = float32 this.Q
        let y = float32 this.R
        Vector2(x, y)

    member self.MoveQ length =
        Hexagon.NewCube(self.Q + length) self.R (self.S - length)

    member self.MoveR length =
        Hexagon.NewCube self.Q (self.R + length) (self.S - length)

    member self.MoveS length =
        Hexagon.NewCube(self.Q - length) (self.R + length) self.S

    member self.DistanceTo(other: Hexagon) =
        // https://www.redblobgames.com/grids/hexagons/#distances-cube
        ((abs <| (self.Q - other.Q))
         + (abs <| (self.R - other.R))
         + (abs <| (self.S - other.S)))
        / 2

    member self.IsNeighbour other = self.DistanceTo other = 1

    member self.GetNeighbour direction =
        match direction with
        | Direction.East -> Hexagon.NewCube(self.Q + 1) self.R (self.S - 1)
        | Direction.NorthEast -> Hexagon.NewCube(self.Q + 1) (self.R - 1) self.S
        | Direction.NorthWest -> Hexagon.NewCube self.Q (self.R - 1) (self.S + 1)
        | Direction.West -> Hexagon.NewCube(self.Q - 1) self.R (self.S + 1)
        | Direction.SouthWest -> Hexagon.NewCube(self.Q - 1) (self.R + 1) self.S
        | Direction.SouthEast -> Hexagon.NewCube self.Q (self.R + 1) (self.S - 1)
        | _ -> raise (ArgumentOutOfRangeException("direction"))

File: Strategy.Game/Hexagon.cs
using Microsoft.Xna.Framework;

namespace Strategy.Game;

public record struct Hexagon(int Q, int R, int S)
{
    public enum Direction
    {
        East,
        NorthEast,
        NorthWest,
        West,
        SouthWest,
        SouthEast,
    }

    private static (int rx, int ry, int rz) CubeRound(float x, float y, float z)
    {
        float rx = MathF.Round(x);
        float ry = MathF.Round(y);
        float rz = MathF.Round(z);

        float xDiff = Math.Abs(rx - x);
        float yDiff = Math.Abs(ry - y);
        float zDiff = Math.Abs(rz - z);

        if (xDiff > yDiff && xDiff > zDiff)
        {
            rx = -ry - rz;
        }
        else if (yDiff > zDiff)
        {
            ry = -rx - rz;
        }
        else
        {
            rz = -rx - ry;
        }

        return ((int)rx, (int)ry, (int)rz);
    }

    private static int CalculateAxis(int axis1, int axis2) => -axis1 - axis2;

    public static Hexagon Zero => new Hexagon(0, 0, 0);

    public static Hexagon NewAxial(int q, int r) => new(q, r, CalculateAxis(q, r));

    public static Hexagon At2DPosition(Vector2 position, float cellSize)
    {
        float q =
            (MathF.Sqrt(3f) / 3f * position.X - 1f / 3f * position.Y)
            / cellSize;

        float r = (2f / 3f * position.Y) / cellSize;
        float s = -q - r;
        (int rx, int ry, int rz) = CubeRound(q, r, s);
        return new Hexagon(rx, ry, rz);
    }
    
    public Vector2 Get2DPosition(float cellSize)
    {
        float x =
            cellSize
            * (MathF.Sqrt(3f) * Q
            + MathF.Sqrt(3f) / 2f * R);

        var y = cellSize * (3f / 2f * R);

        return new Vector2(x, y);
    }

    public Hexagon FromVector2(Vector2 vector) => NewAxial((int)vector.X, (int)vector.Y);

    public Hexagon MoveQ(int length) => new(Q + length, R, S - length);
    public Hexagon MoveR(int length) => new(Q, R + length, S - length);
    public Hexagon MoveS(int length) => new(Q - length, R, S + length);

    public int DistanceTo(Hexagon other)
    {
        // https://www.redblobgames.com/grids/hexagons/#distances-cube
        return (Math.Abs(Q - other.Q) + Math.Abs(R - other.R) + Math.Abs(S - other.S)) / 2;
    }

    public bool IsNeighbor(Hexagon other) => this.DistanceTo(other) == 1;

    public Hexagon GetNeighbor(Direction direction)
    {
        return direction switch
        {
            Direction.East => new Hexagon(Q + 1, R, S - 1),
            Direction.NorthEast => new Hexagon(Q + 1, R - 1, S),
            Direction.NorthWest => new Hexagon(Q, R - 1, S + 1),
            Direction.West => new Hexagon(Q - 1, R, S + 1),
            Direction.SouthWest => new Hexagon(Q - 1, R + 1, S),
            Direction.SouthEast => new Hexagon(Q, R + 1, S - 1),
            _ => throw new ArgumentOutOfRangeException(nameof(direction), direction, null),
        };
    }
}
